src/main/java/spoon/reflect/visitor/ImportAnalyzer.java

```diff
@@ -0,0 +1,236 @@
+		}
+		scanner.exit(cu);
+	}
+
+	protected CtScannerListener createScannerListener(T scanner) {
+		return new ScannerListener(scanner);
+	}
+
+	//The set of roles whose values are always kept implicit
+	protected static Set<CtRole> IGNORED_ROLES_WHEN_IMPLICIT = new HashSet<>(Arrays.asList(
+			//e.g. List<String> s = new ArrayList</*keep me implicit*/>();
+			CtRole.TYPE_ARGUMENT,
+			//e.g. List<?/* extends Object*/>
+			CtRole.BOUNDING_TYPE,
+			//e.g. (/*implicit type of parameter*/ p) -> {}
+			CtRole.TYPE
+	));
+
+	/**
+	 * {@link CtScannerListener} implementation which stops scanning of children on elements,
+	 * which mustn't have influence to compilation unit imports.
+	 */
+	protected class ScannerListener implements CtScannerListener {
+		protected T scanner;
+		protected Set<CtRole> ignoredRoles = IGNORED_ROLES_WHEN_IMPLICIT;
+
+		ScannerListener(T scanner) {
+			super();
+			this.scanner = scanner;
+		}
+
+		@Override
+		public ScanningMode enter(CtRole role, CtElement element) {
+			if (element == null) {
+				return ScanningMode.SKIP_ALL;
+			}
+			if (role == CtRole.VARIABLE && element instanceof CtVariableReference) {
+				//ignore variable reference of field access. The accessType is relevant here instead.
+				return ScanningMode.SKIP_ALL;
+			}
+			if (element.isParentInitialized()) {
+				CtElement parent = element.getParent();
+				if (role == CtRole.DECLARING_TYPE && element instanceof CtTypeReference) {
+					if (parent instanceof CtFieldReference) {
+						//ignore the declaring type of field reference. It is not relevant for Imports
+						return ScanningMode.SKIP_ALL;
+					}
+					if (parent instanceof CtExecutableReference) {
+						/*
+						 * ignore the declaring type of type executable like
+						 * anVariable.getSomeInstance().callMethod()
+						 * The declaring type of `callMethod` method is not relevant for Imports
+						 */
+						return ScanningMode.SKIP_ALL;
+					} else if (parent instanceof CtTypeReference) {
+						/*
+						 * It looks like this is not needed too.
+						 *
+						 * pvojtechovsky: I am sure it is not wanted in case of
+						 * spoon.test.imports.testclasses.internal.ChildClass.InnerClassProtected
+						 * which extends package protected (and for others invisible class)
+						 * spoon.test.imports.testclasses.internal.SuperClass
+						 * and in this case the import directive must import ...ChildClass and not ...SuperClass,
+						 * because import is using type "access path" and not qualified name of the type.
+						 *
+						 * ... but in other normal cases, I guess the declaring type is used and needed for import!
+						 * ... so I don't understand why SKIP_ALL works in all cases. May be there is missing test case?
+						 */
+						if (!((CtTypeReference) parent).getAccessType().equals(element)) {
+							return ScanningMode.SKIP_ALL;
+						}
+					} else {
+						//May be this can never happen
+						throw new SpoonException("Check this case. Is it relvant or not?");
```

Comment:
Typo: relvant -> relevant

Reply:
thanks @seintur but it's already  merged, would you edit the file directly on Github?